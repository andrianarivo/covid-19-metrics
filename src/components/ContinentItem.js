import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';
import styled from 'styled-components';
import moment from 'moment';
import africa from '../images/africa.png';
import europe from '../images/europe.png';
import northAmerica from '../images/north_america.png';
import asia from '../images/asia.png';
import oceania from '../images/oceania.png';
import southAmerica from '../images/south_america.png';

const getImage = (continent) => {
  switch (continent.toLowerCase()) {
    case 'asia':
      return asia;
    case 'africa':
      return africa;
    case 'north america':
      return northAmerica;
    case 'south america':
      return southAmerica;
    case 'australia-oceania':
      return oceania;
    case 'europe':
      return europe;
    default:
      return asia;
  }
};

export default function ContinentItem(props) {
  const navigate = useNavigate();
  const { continent, active, updated } = props;

  return (
    <Continent image={continent}>
      <p>{continent}</p>
      <p>
        <span>{active}</span>
        {' '}
        <span>active cases</span>
      </p>
      <p>
        updated
        {` ${moment(new Date(updated)).fromNow()}`}
      </p>
      <button
        type="button"
        data-testid={`button-${continent}`}
        onClick={() => {
          navigate(`details/${continent}`);
        }}
      >
        {' '}
      </button>
    </Continent>
  );
}

ContinentItem.propTypes = {
  continent: PropTypes.string.isRequired,
  active: PropTypes.number.isRequired,
  updated: PropTypes.number.isRequired,
};

const Continent = styled.li`
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  justify-content: flex-end;
  position: relative;
  list-style: none;
  flex: 1 1 50%;
  padding: 1rem 0;
  height: 10rem;
  background-image: url(${(props) => getImage(props.image)});
  background-size: contain;
  background-repeat: no-repeat;
  filter: blur(5);

  button {
    position: absolute;
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;
    background-color: transparent;
    border: none;
    cursor: pointer;
  }

  p {
    padding-right: 0.5rem;

    span:nth-child(2) {
      font-size: 0.7rem;
    }
  }

  p:nth-child(1) {
    font-weight: bold;
    text-transform: uppercase;
    text-align: end;
  }

  p:nth-child(3) {
    font-size: 0.7rem;
  }

  &:nth-child(n) {
    background-color: #dc4781;
  }

  &:nth-child(3n) {
    background-color: #cf4378;
  }

  &:nth-child(2) {
    background-color: #cf4378;
  }
`;
