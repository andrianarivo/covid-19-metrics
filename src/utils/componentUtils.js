import asia from '../images/asia.png';
import africa from '../images/africa.png';
import northAmerica from '../images/north_america.png';
import southAmerica from '../images/south_america.png';
import oceania from '../images/oceania.png';
import europe from '../images/europe.png';

const getImage = (continent) => {
  switch (continent.toLowerCase()) {
    case 'asia':
      return asia;
    case 'africa':
      return africa;
    case 'north america':
      return northAmerica;
    case 'south america':
      return southAmerica;
    case 'australia-oceania':
      return oceania;
    case 'europe':
      return europe;
    default:
      return asia;
  }
};

export default getImage;
