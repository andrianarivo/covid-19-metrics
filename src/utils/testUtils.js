import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../redux/store';

export function withRouter(children) {
  return (
    <BrowserRouter>
      {children}
    </BrowserRouter>
  );
}

export function withProvider(children, mockStore = store) {
  return (
    <Provider store={mockStore}>
      {children}
    </Provider>
  );
}
