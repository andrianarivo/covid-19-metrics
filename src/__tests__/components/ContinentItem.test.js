import { create } from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ContinentItem from '../../components/ContinentItem';
import { withRouter } from '../../utils/testUtils';

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockUseNavigate,
}));

describe('Testing ContinentItem component', () => {
  test('renders correctly', () => {
    const tree = create(withRouter(<ContinentItem
      continent="Dummy"
      active={0}
      updated={0}
    />));
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test('clicks the button and navigate', async () => {
    // Arrange
    render(<ContinentItem continent="Dummy" active={0} updated={0} />);

    // Act
    const button = screen.getByTestId('button-Dummy');
    await userEvent.click(button);

    // Assert
    expect(mockUseNavigate).toBeCalledWith('details/Dummy', {
      state: {
        active: 0,
        continent: 'Dummy',
        updated: 0,
      },
    });
    mockUseNavigate.mockRestore();
  });
});
