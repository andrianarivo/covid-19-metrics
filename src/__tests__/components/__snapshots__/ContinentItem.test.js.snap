// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ContinentItem component renders correctly 1`] = `
<li
  className="sc-kIDzkD jHRNaF"
  image="Dummy"
>
  <p>
    Dummy
  </p>
  <p>
    0
  </p>
  <p>
    53 years ago
  </p>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-circle-arrow-right sc-jzuGiw knwkHW"
    data-icon="circle-arrow-right"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={Object {}}
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 256a256 256 0 1 0 512 0A256 256 0 1 0 0 256zM281 385c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l71-71L136 280c-13.3 0-24-10.7-24-24s10.7-24 24-24l182.1 0-71-71c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0L393 239c9.4 9.4 9.4 24.6 0 33.9L281 385z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <button
    data-testid="button-Dummy"
    onClick={[Function]}
    type="button"
  >
     
  </button>
</li>
`;
